### ISSUE TITLE
Communication between machines and the world outside the MIO event loop

### ISSUE THREAD

**Issue by blabaere** [tailhook/rotor#6]
Hello,

After reading the 3rd post about rotor, I had a quick look to check if it could fit my needs (rust implementation of nanomsg's scalability protocols) and it seems to be quite close. The only thing that I feel is missing is a way for a client outside the event loop to talk with the machines. 

Modifying the tcp_echo_server as follow, one can manage to send a single message to the machine, but only once.

``` rust
    let mut port : Option<Port<u8>> = None;
    let lst = TcpListener::bind(&"127.0.0.1:3000".parse().unwrap()).unwrap();
    let ok = handler.add_machine_with(&mut event_loop, |scope| {
        let (p, _) = scope.create_future();
        port = Some(p);
        Ok(Echo::new(lst, scope))
    }).is_ok();
    assert!(ok);
    event_loop.run(&mut handler).unwrap();

    if let Some(p) = port.take() {
        p.set(42);
    }
```

The way the event loop channel is hidden behind a oneshot port/future pair makes it impossible (as far as I can see) to send an arbitrary number of messages to the machine. Is there a way to send several messages that I didn't find out ?

Regards,
Benoît

**Comment by tailhook** [comment]
Hi Benoît,

Yes, the Future/Port combination is limited. It should be possible to consume future, then send Port somewhere again for the next message. While it may work for some cases (like communicating to the pool of workers). It looks like too ugly, for many of them.

I have used to think that I could implement some super-trivial communication primitive in rotor itself and allow to build more primitives on top. "Future" is such a basic abstraction in python asyncio, and looks like similar to the Promise, which is the same for JavaScript. But it's not like that for Rust I think.

There are couple of alternatives though. I'm not sure which one (or multiple things) should be included in the core.

Do you have any ideas?

BTW, I'm interested in nanomsg protocols too, please keep me informed :)

**Comment by blabaere** [comment]
Thanks for the super fast answer.

The first thing that comes to mind would be to make the Notify and Timeo enums generic :

``` rust
pub enum Notify<T:Send> {
    Fsm(Token, T),
}

pub enum Timeo<T:Send> {
    Fsm(Token, T),
}
```

This would allow the client to use directly the MIO event loop channel, with the problem of adding two generic parameters almost everywhere (Handler, Scope etc). But then, to be composable, the machines would have to share the same Notify and Timeo generic type.

**Comment by blabaere** [comment]
Regarding nanomsg rust implementation, the first rotor machine would be the equivalent of the zmq context, sockets would also be machines, and finally pipes would also be machines, performing the I/O work. In that design, some user facing structs (the zmq context and sockets) would send through the MIO event loop channel commands like 'create socket', 'connect', 'bind', 'set_option', 'send', 'recv' and a regular rust channel would be used by the machines to send the result of the commands back to the user facing structs.

So this is where something similar to a channel rather than a port/future pair would be required.
You can check the existing attempt [here](https://github.com/blabaere/scaproust), but be warned: this is far from being as clean and readable as the rotor code.

**Comment by tailhook** [comment]
> the first rotor machine would be the equivalent of the zmq context, sockets would also be machines, and finally pipes would also be machines

While this design might be okay. I want to propose an alternative.

You may put a nanomsg context into a Context. Like [this one](https://github.com/tailhook/rotor-http/blob/master/src/server/context.rs) or [this one](https://github.com/tailhook/rotor-http/blob/master/src/client/context.rs) (the former is okay for settings, latter is useful if you want to store state there).

Socket story is a little bit more complicated:
1. Pull/Sub socket is just a state machine that receives a messages
2. Push/Pub sockets could be `Arc<Mutex<>>`-wrapped structure that allows you to send data directly into right pipe(s)
3. Rep socket should probably spawn a state machine on every request, so you can process requests simultaneously
4. Req socket is very similar to Push, but should return Future, that will receive a response

Just my 2cents, not sure if all that makes sense.

**Comment by tailhook** [comment]
@blabaere 

Okay. I have a little progress. Here is a "monitor" branch: https://github.com/tailhook/rotor/tree/monitor

Here is [a test to demonstrate how does it work](https://github.com/tailhook/rotor/blob/monitor/src/monitor.rs#L227). The basic idea is that you create a pair of objects, each of them implement `Monitor<T>` trait. You can use it mostly as a bidirectional pipe by using `Monitor::notify_peer` and `Monitor::consume` methods. Except there is no actual pipe or queue. There is a type `T` which is accesible by `Monitor::consume` method.

While you can put queue(s) in `T`, there are more efficient patterns here. For example, you could put an actual socket to the T, and let other peer to write to the socket directly (sure, this should be encapsulated by carefully designed state machine).

The Future/Port thing is going to be deprecated.

This is just brief description of the abstraction. I'm going to write another article that describes the design in more comprehensive way. But it would be great if you could take an early look and let me know if that makes sense or not.

/cc @erikjohnston

**Comment by blabaere** [comment]
I feel 'a bidirectional pipe' is really what is needed here, so I guess it is going the right direction.
After reading the example I tried to modify tcp_echo_server to use monitors, and couldn't get it to do what I wanted. To be a little bit more explicit, what I expect is something that would send values to the event loop channel, like it is currently doing, and another kind of channel for the 'inside to outside' communication. That is, with both blocking and non-blocking read functions.

The example modified:

``` rust
extern crate mio;
extern crate rotor;

use std::io::{Write, stderr};
use std::error::Error;

use mio::{EventSet, PollOpt, TryRead, TryWrite};
use mio::tcp::{TcpListener, TcpStream};
use rotor::{Machine, Response, Scope, Monitor, Peer1Monitor};

struct Context;

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
enum CmdAndEvt {
    Boot,
    Cmd(usize),
    Evt(usize)
}

enum Echo {
    Server(TcpListener, Peer1Monitor<CmdAndEvt>),
    Connection(TcpStream),
}

impl Echo {
    pub fn new(
        sock: TcpListener, 
        scope: &mut Scope<Context>, 
        mon: Peer1Monitor<CmdAndEvt>) -> Echo {
        scope.register(&sock, EventSet::readable(), PollOpt::edge()).unwrap();
        Echo::Server(sock, mon)
    }
    fn accept(self) -> Response<Echo, TcpStream> {
        match self {
            Echo::Server(sock, mon) => {
                match sock.accept() {
                    Ok(Some((conn, _))) => {
                        Response::spawn(Echo::Server(sock, mon), conn)
                    }
                    Ok(None) => {
                        Response::ok(Echo::Server(sock, mon))
                    }
                    Err(e) => {
                        writeln!(&mut stderr(), "Error: {}", e).ok();
                        Response::ok(Echo::Server(sock, mon))
                    }
                }
            }
            _ => unreachable!(),
        }
    }
}

impl Machine<Context> for Echo {
    type Seed = TcpStream;

    fn create(conn: TcpStream, scope: &mut Scope<Context>)
        -> Result<Self, Box<Error>>
    {
        scope.register(&conn, EventSet::readable(), PollOpt::level())
            .unwrap();
        Ok(Echo::Connection(conn))
    }

    fn ready(self, _events: EventSet, _scope: &mut Scope<Context>)
        -> Response<Self, TcpStream>
    {
        match self {
            me @ Echo::Server(..) => me.accept(),
            Echo::Connection(mut sock) => {
                let mut data = [0u8; 1024];
                match sock.try_read(&mut data) {
                    Err(e) => {
                        writeln!(&mut stderr(), "read: {}", e).ok();
                        Response::done()
                    }
                    Ok(Some(0)) => {
                        Response::done()
                    }
                    Ok(Some(x)) => {
                        match sock.try_write(&data[..x]) {
                            Ok(_) => {
                                // this is example so we don't care if not all
                                // (or none at all) bytes are written
                                Response::ok(Echo::Connection(sock))
                            }
                            Err(e) => {
                                writeln!(&mut stderr(), "write: {}", e).ok();
                                Response::done()
                            }
                        }
                    }
                    Ok(None) => {
                        Response::ok(Echo::Connection(sock))
                    }
                }
            }
        }
    }
    fn spawned(self, _scope: &mut Scope<Context>) -> Response<Self, TcpStream>
    {
        match self {
            me @ Echo::Server(..) => me.accept(),
            _ => unreachable!(),
        }
    }
    fn timeout(self, _scope: &mut Scope<Context>)
        -> Response<Self, TcpStream>
    {
        unreachable!();
    }
    fn wakeup(self, _scope: &mut Scope<Context>)
        -> Response<Self, TcpStream>
    {
        match self {
            Echo::Server(sock, mon) => {
                match mon.consume() {
                    Some(mut item) => {
                        println!("Echo::wakeup, consumed: {:?} !", *item);
                        *item = CmdAndEvt::Evt(666);
                    },
                    None => {
                        println!("Echo::wakeup, consume failed !");
                    }
                }

                Response::ok(Echo::Server(sock, mon))
            }
            _ => unreachable!()
        }
    }
}

fn main() {
    let mut event_loop = mio::EventLoop::new().unwrap();
    let mut handler = rotor::Handler::new(Context, &mut event_loop);
    let lst = TcpListener::bind(&"127.0.0.1:3000".parse().unwrap()).unwrap();
    let mut outer_pipe_end = None;
    let ok = handler.add_machine_with(&mut event_loop, |scope| {
        let (left_end, right_tok) = scope.create_monitor(CmdAndEvt::Boot);
        let right_end = match right_tok.connect(scope) {
            Ok(x) => Some(x),
            Err(_) => None
        };

        outer_pipe_end = right_end;
        Ok(Echo::new(lst, scope, left_end))
    }).is_ok();
    assert!(ok);
    println!("Waiting for the event loop to wake up ...");
    std::thread::spawn(move || event_loop.run(&mut handler).unwrap());
    std::thread::sleep(std::time::Duration::from_millis(500));
    println!("Finished waiting.");
    if let Some(pipe) = outer_pipe_end.take() {
        println!("outer_pipe_end is initialized.");

        loop {
            match pipe.consume() {
                Some(mut item) => {
                    println!("Read from pipe: {:?}", *item);
                    *item = CmdAndEvt::Cmd(42);
                },
                None => {
                    println!("Read from pipe failed !");
                    std::thread::sleep(std::time::Duration::from_millis(500));
                },
            }
        }
    } else {
        println!("outer_pipe_end is NOT initialized !");
    }
}
```

And the output:

```
Waiting for the event loop to wake up ...
Echo::wakeup, consumed: Boot !
Finished waiting.
outer_pipe_end is initialized.
Read from pipe failed !
Read from pipe failed !
```

**Comment by tailhook** [comment]
Okay. Here is a fixed example:
https://gist.github.com/anonymous/452cbe6437256edeca9f

(sorry, I forgot to make `notify_peer` method public)

The blocking communication can be implemented via [Condvar](https://doc.rust-lang.org/stable/std/sync/struct.Condvar.html). The monitor abstraction is a thing similar to `Arc<(Mutex<T>, Condvar)>` in the documentation example. Except its notify_peer function wakes up a state machine inside the event loop. And there are actually two conditional variables inside. Hopefully, this makes things a bit more clear.

API is a little bit confusing at the moment, I'll try to fix it soon.

**Comment by tailhook** [comment]
Well, after some thinking I've simplified it even more. Now we have
`scope.notifier()` that returns `Notifier` object. And `notifier.wakeup()` just wakes up a state machine (asynchronously) which initially called `scope.notifier()`. This is it. The notifier is `Send+Sync+Clone`.

This should be good enough to build abstractions on top. Note however, that spurious wakeups are possible and entirely normal on this level of abstraction.

The code is in current master along with few more upcoming changes (like removing futures and making context an associated type). I'm working on updating all the dependencies.

Let me know what you think.

**Comment by blabaere** [comment]
I was about to write this:

_I think this is the simplest feature possible, as low level as it can get and puts no additional constraint on rotor design. With this and the synchronization primitives availabe in Rust std (channel, or Arc + Mutex + Condvar) it should also be possible to build whatever one may need._

And then I noticed that `Notifier::wakeup` is consuming the notifier. I did a quick check and changing the signature from `Notifier::wakeup(self)` to `Notifier::wakeup(&mut self)` would make it what's described above ! What's the idea behind making the notifier usable only once ? Did I miss something ?

**Comment by tailhook** [comment]
@blabaere sorry, that was a mistake. Pushed new version.

**Comment by tailhook** [comment]
That was released in rotor 0.5.0. Looks fine so far.
